src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs

```diff
@@ -94,28 +114,29 @@ protected internal XmlDocument(XmlImplementation imp) : base()
             _implementation = imp;
             _domNameTable = new DomNameTable(this);
 
-            // force the following string instances to be default in the nametable
-            XmlNameTable nt = this.NameTable;
-            nt.Add(string.Empty);
-            strDocumentName = nt.Add("#document");
-            strDocumentFragmentName = nt.Add("#document-fragment");
-            strCommentName = nt.Add("#comment");
-            strTextName = nt.Add("#text");
-            strCDataSectionName = nt.Add("#cdata-section");
-            strEntityName = nt.Add("#entity");
-            strID = nt.Add("id");
-            strNonSignificantWhitespaceName = nt.Add("#whitespace");
-            strSignificantWhitespaceName = nt.Add("#significant-whitespace");
-            strXmlns = nt.Add("xmlns");
-            strXml = nt.Add("xml");
-            strSpace = nt.Add("space");
-            strLang = nt.Add("lang");
-            strReservedXmlns = nt.Add(XmlReservedNs.NsXmlNs);
-            strReservedXml = nt.Add(XmlReservedNs.NsXml);
-            strEmpty = nt.Add(String.Empty);
+            strXmlns = Xmlns;
+            strXml = Xml;
+            strReservedXmlns = XmlReservedNs.NsXmlNs;
+            strReservedXml = XmlReservedNs.NsXml;
             baseURI = String.Empty;
-
             objLock = new object();
+
+            NameTable nt = imp.NameTable as NameTable;
+            if (nt != null)
```

Comment:
Presumably this is in part to avoid compat issues, since we were previously calling Add and that could have had custom logic?  I think it'd be worth a comment describing why the two different code paths here.